Add SEP defining transactions required (#3)
stellar/experimental-payment-channels: specifications/sep-payment-channel-mechanism.md

have been exceeded. These two properties together are referred to as O
throughout the protocol.

The participants may agree to change the period O at anytime by following the
[Change the Observation Period](#Change-the-Observation-Period) process.

### Accounts

The payment channel utilizes two Stellar accounts that are both 2-of-2
multisig accounts while the channel is open:

- E, the _escrow account_, that holds the assets that both participants have
contributed to the channel and that will be distributed to the participants
at channel close according to the final close transactions submitted.
Created by I.  Jointly controlled by I and R while the channel is open.
Control is returned to I at close.

- V, the _reserve account_, that holds an amount of native asset contributed
by R that will be used to sponsor the claimable balance ledger entries
created at disbursement.  Created by R.  Jointly controlled by I and R while
the channel is open.  Control is returned to R at close.  Cannot be merged
until all claimable balances created at close for R are claimed by R.

### Constants

The two participants agree on the following constants:

> - m, the _maximum transaction count for an iteration's transaction set_, is

Author: I think the definition of m is worthwhile because otherwise we have magic numbers, in this case 2, throughout the formulas. That was something I found difficult to understand in the original specification of the protocol in CAP-21.

---

I commented throughout about why I don't think it is suitable to use m in the other places.

> we should potentially replace `1` with `m`

Note that m has value 2.

> - m, the _maximum transaction count for an iteration's transaction set_, is

Reviewer: Hm yeah I agree that's better than magic numbers. Makes sense to me!

sequence number of escrow account E after E has been created. It is the first
available sequence number for iterations to consume.

- i, the _iteration number_, is initialized to zero.  It is incremented with
every off-chain update of the payment channel state, or on-chain setup, deposit,
withdrawal, etc.

- e, the _executed iteration number_, is initialized to zero. It is incremented
with every iteration submitted on-chain.

### Computed Values

The two participants frequently use the following computed values:

- s_i, the _iteration sequence number_, is the sequence number that iteration
i's transaction set starts at. Assuming the history of the payment channel has a
single value for m it is computable as, s+(m*i).

- s_e, the _executed iteration sequence number_, is the sequence number that the
executed iteration e's transaction set starts at. Assuming the history of the
payment channel has a single value for m it is computable as, s+(m*e).

### Processes

#### Setup

To setup the payment channel:

1. I creates the escrow account E.
2. R creates the reserve account V.
3. Set variable initial states:
   - s to E's sequence number + 1.
   - i to 0.

>    - e to 0.

Author: I don't think so, because it's not a variable, it's a constant with a predefined value.

    signers to E.
    - One or more `SET_OPTIONS` operations adding I's signers to V.
    - One or more `CHANGE_TRUST` operations adding trustlines to E.
    - One `END_SPONSORING_FUTURE_RESERVES` operation that stops I sponsoring
    future reserves of subsequent operations.
  - Operations sponsored by R:
    - One `BEGIN_SPONSORING_FUTURE_RESERVES` operation that specifies reserve
    account R as a sponsor of future reserves.
    - One or more `SET_OPTIONS` operations adjusting escrow account V's
    thresholds such that R and I's signers must both sign, and adding R's
    signers to V.
    - One or more `SET_OPTIONS` operations adding R's signers to E.
    - One `END_SPONSORING_FUTURE_RESERVES` operation that stops R sponsoring
    future reserves of subsequent operations.
  - One or more `PAYMENT` operations depositing I's contribution to E.
  - One or more `PAYMENT` operations depositing R's contribution to E.
  - One or more `PAYMENT` operations depositing R's reserves to V, for each
  trustline on E that will be used to sponsor claimable balances at
  disbursement.
  
- C_i, see [Update](#Update) process.

- D_i, see [Update](#Update) process.

> #### Update

Reviewer: A general note about this section: one of the main reasons for updating a payment channel is to effectively make an off-chain payment between the two participants. 

This is clearly enabled by the mechanics described below (if I correctly understood you simply have to update the value on the `CREATE_CLAIMABLE_BALANCE` operation), but maybe it's worth making this more explicit? After all, I would guess this is the main functionality people are looking for when implementing this protocol.

Author: Good call. Added in d653b78.

Reviewer: Looks great!

To update the payment channel state, the participants:

1. Increment i.
2. Sign and exchange a closing transaction C_i.
3. Sign and exchange a declaration transaction D_i.

The transactions are constructed as follows:

- C_i, the _closing transaction_, disburses funds to R and changes the signing
weights on E such that I unilaterally controls E.  C_i has source account E,

> sequence number s_i+1, a `minSeqAge` of O (the observation period time

Reviewer: Wouldn't this be `s_i+m`?

Author: The C tx needs to be the second tx in the tx set. s_i is the seq of the first tx in the tx set, that's why C is s_i+1.

  The `minSeqAge` and `minSeqLedgerGap` prevents a misbehaving party from
  executing C_i when the channel state has already progressed to a later
  iteration number, as the other party has the period O to invalidate C_i by
  submitting D_i' for some i' > i.
  
  C_i contains operations:
  - One `BEGIN_SPONSORING_FUTURE_RESERVES` operation that specifies reserve
  account V as a sponsor of future reserves.
  - One `CREATE_CLAIMABLE_BALANCE` operation for each trustline that is
  disbursing funds to R.
  - One `END_SPONSORING_FUTURE_RESERVES` operation that confirms reserve
  account V's sponsorship.
  - One or more `SET_OPTIONS` operation adjusting escrow account E's
  thresholds to give I full control of E, and removing R's signers.
  - One or more `SET_OPTIONS` operation adjusting reserve account E's
  thresholds to give R full control of V, and removing I's signers.

- D_i, the _declaration transaction_, declares an intent to execute the
corresponding closing transaction C_i.  D_i has source account E, sequence
number s_i, and `minSeqNum` set to s_e.  Hence, D_i can execute at any time, so
long as E's sequence number n satisfies s_e <= n < s_i.  Because C_i has source

> account E and sequence number s_i+1, D_i leaves E in a state where C_i can

Reviewer: Same as above: `s_i+m`

transaction.

1. Submit most recent D_i
2. Wait observation period O
3. Submit C_i

#### Contesting a Close

Participants can contest a close if the close is not the most recent agreed
closing state of the payment channel.

Participants can attempt to close the channel at a state that is earlier in the
history of the channel than the most recently agreed to state. A participant who
is a malicious actor might attempt to do this if an earlier state benefits them.

The malicious participant can do this by performing the [Uncoordinated
Close](#Uncoordinated-Close) process with a declaration transaction that is not
the most recently signed declaration transaction.

The other participant can identify that the close process has started at an
earlier state by monitoring changes in escrow account E's sequence. If the other
participant sees the sequence number of escrow account E change to a value that
is not the most recently used s_i, they can use the following process to contest
the close. A participant contests a close by submitting a more recent
declaration transaction and closing the channel at the actual final state.

1. Get E's sequence number n

> 2. If s_{e+1} >= n < s_i, go to step 3, else go to step 1

Reviewer: `s_{e+m}`

Author: If we wanted to use m here it would be s_e+m, but the intention is to communicate the seq for the next iteration after e, which I think is clearer as it is. If m ever changed too it means the boosting here doesn't have to deal with that.

3. Submit most recent D_i
4. Wait observation period O

> 5. Submit C_i

Reviewer: Maybe it's worth adding a paragraph explaining why this works? Effectively you're submitting a transaction that bumps the sequence number to `s_i` which is higher than `s_{e+n}`, which invalidates the transaction the malicious participant submitted.

Author: Great call. I don't say why this works here at all. 24543b0

Reviewer: Perfect, thanks!

`minSeqNum` is also set in the future. Participants then sign a transaction to
make the change that only moves the sequence of escrow account E to satisfy the
`minSeqNum` of the future D_i.

Operations that can fail and change the balances of the channel have the
following requirements as well:

- The transaction that can fail must have its source account set to an account
that is not escrow account E.
- The transaction that can fail must contain a `BUMP_SEQUENCE` operation that
bumps escrow account E's sequence number to a sequence number that makes the D_i
executable.

Operations where failure cannot occur or is of no consequence:

- [Change the Observation Period](#Change-the-Observation-Period)
- [Add Trustline](#Add-Trustline)
- [Remove Trustline](#Remove-Trustline)

Operations that can fail:
- [Deposit by Initiator](#Deposit-by-Initiator)
- [Deposit by Responder](#Deposit-by-Responder)
- [Withdraw](#Withdraw)

##### Add Trustline

Participants can add additional trustlines if they plan to make deposits of new balances.

1. I and R sign and exchange signatures for trustline transaction TA_i.
2. I or R submit TA_i.

> If the remove trustline transaction TA_i fails or is never submitted, there is

Reviewer: You probably meant **add** trustline here?

Author: Thanks! d4022e5

  trustline being removed from E that would have been used to sponsor claimable
  balances at disbursement and are no longer required.

##### Deposit by Initiator

Participant I may deposit into the channel without coordination with
participant R, as long as escrow account E already has a trustline for the
asset being deposited.

If participant I wishes to deposit an asset that escrow account E does not hold
a trustline for, the [Add Trustlines](#Add-Trustline) process must be used
first.

##### Deposit by Responder

Participant R may deposit into the channel without coordination with participant
I, as long as escrow account E already has a trustline for the asset being
deposited, and as long as participants R's intent is to make a payment of the
same value to participant I. Any amounts deposited to the payment channel
without coordination will be disbursable to participant I at close.

Participant R must coordinate with participant I to deposit any amount that it
does not intend to immediately pay participant I. The participants use the
following process:

1. Increment i.
2. I and R build the deposit transaction P_i.

> 3. Set e' to e.

Reviewer: It took me a while to realize `e'` is an auxiliary var you're using, since all the other vars are well-defined concepts here. Maybe change the phrasing to `Store the current value for e on e'`?

Author: ```suggestion
3. Set e' to the value of e.
```

Author: Ah yes this is not particularly clear. Thanks. I used slightly different language in ff8ec83.

7. Sign and exchange a declaration transaction D_i.
8. I and R sign and exchange signatures for deposit transaction P_i.
9. I or R submit P.

If the deposit transaction P fails or is never submitted, the C_i and D_i are
not executable because escrow account E's sequence number was not bumped to s_i.
The participants should take the following steps since the deposit did not
succeed:

10. Set e to e'.

The transactions are constructed as follows:

- P_i, the _deposit transaction_, makes one or more payments from any Stellar
accounts to escrow account E. P_i has any source account that is not E or V,
typically the participant proposing the change.

  P_i contains operations:

  - One or more `PAYMENT` operations depositing assets into escrow account E.
  - One `BUMP_SEQUENCE` operation bumping the sequence number of escrow account
  E to s_i.
  
- C_i, see [Update](#Update) process.

- D_i, see [Update](#Update) process.

##### Withdraw

Participants must coordinate to withdraw an amount without closing the channel.
The participants use the following process, where W is the participant
withdrawing:

1. Increment i.
2. I and R build the withdrawal transaction W_i.

> 3. Set e' to e.

Reviewer: Same comment about `e'` as above.

Author: ff8ec83

through claimable balances at channel close.

### Observation Period

A payment channel defines an observation period O within which all
participants are guaranteed to be able to observe the blockchain state and
submit transactions in response to changing state.

The participants agree on the period O at channel open.

The observation period is defined both as a duration in time, and a count of
ledgers. The observation period has passed if both the duration and ledger count
have been exceeded. These two properties together are referred to as O
throughout the protocol.

The participants may agree to change the period O at anytime by following the
[Change the Observation Period](#Change-the-Observation-Period) process.

### Accounts

The payment channel utilizes two Stellar accounts that are both 2-of-2
multisig accounts while the channel is open:

- E, the _escrow account_, that holds the assets that both participants have
contributed to the channel and that will be distributed to the participants
at channel close according to the final close transactions submitted.
Created by I.  Jointly controlled by I and R while the channel is open.
Control is returned to I at close.

> - V, the _reserve account_, that holds an amount of native asset contributed

Reviewer 2: a second multisig account eventually owned by R for R's claimable balances makes sense to me

by R that will be used to sponsor the claimable balance ledger entries
created at disbursement.  Created by R.  Jointly controlled by I and R while
the channel is open.  Control is returned to R at close.  Cannot be merged
until all claimable balances created at close for R are claimed by R.

### Constants

The two participants agree on the following constants:

> - m, the _maximum transaction account of an iteration's transaction set_, is

Author: Oops, this is a typo. Should be:
```suggestion
- m, the _maximum transaction count for an iteration's transaction set_, is
```
It's the number of transactions that can be in an iteration, which is yup, always 2.

sequence number s_i+1, a `minSeqAge` of O (the observation period time
duration), and a `minSeqLedgerGap` of O (the observation period ledger count).

  The `minSeqAge` and `minSeqLedgerGap` prevents a misbehaving party from
  executing C_i when the channel state has already progressed to a later
  iteration number, as the other party has the period O to invalidate C_i by
  submitting D_i' for some i' > i.
  
  C_i contains operations:
  - One `BEGIN_SPONSORING_FUTURE_RESERVES` operation that specifies reserve
  account V as a sponsor of future reserves.
  - One `CREATE_CLAIMABLE_BALANCE` operation for each trustline that is
  disbursing funds to R.
  - One `END_SPONSORING_FUTURE_RESERVES` operation that confirms reserve
  account V's sponsorship.
  - One or more `SET_OPTIONS` operation adjusting escrow account E's
  thresholds to give I full control of E, and removing R's signers.
  - One or more `SET_OPTIONS` operation adjusting reserve account E's
  thresholds to give R full control of V, and removing I's signers.

- D_i, the _declaration transaction_, declares an intent to execute the
corresponding closing transaction C_i.  D_i has source account E, sequence

> number s_i, and `minSeqNum` set to s_e.  Hence, D_i can execute at any time, so

Reviewer 2: s_e is a little confusing to me. Say we start at this state:

```
s (E's seqNum) = 1000, i = 0, e = 0
s_i = 1000
s_e = 1000
```

First C_i/D_i get created but not submitted, i =1.
Second C_i/D_i get created and submitted, i = 2, e=1

This second D_i set E.seqNum to 1000 + 2*1 = 1002
This second C_i then sets E.seqNum to 1003

The current state is then:
```
s = 1003, i=2, e=1
s_i = 1004
s_e = 1002
```

So for a third D_i, shouldn't the minSeqNum be 1003, and thus E's current seqNum. And not s_e?

Author: We chatted offline about this, so I think this is resolved. The things that we noted are:
 - Once D is submitted, for e.g. D_2 in the example above, the participants don't sign a third transaction set because the channel is closing.
 - s_e doesn't really matter for channel close, so the doc doesn't mention it being updated in the close steps. We could add it for completeness, but I think it's probably unnecessary? If the second D above was submitted, theoretically e would be equal to 2, the iteration number for the transaction set being submitted.

We used the following table to help think about how s_i and s_e change:
|Description                           |Transaction|Tx Seq|E's seq change|E's seq|s   |i change|i  |e change|e  |s_i |s_e |m  |
|--------------------------------------|-----------|------|--------------|-------|----|--------|---|--------|---|----|----|---|
|Setup: 1. I creates escrow account E  |           |      |1000          |1000   |    |        |   |        |   |    |    |2  |
|Setup: 3. Set variable initial states |           |      |              |1000   |1001|0       |0  |0       |0  |1001|1001|2  |
|                                      |           |      |              |1000   |1001|        |0  |        |0  |1001|1001|2  |
|Setup: 5. Build formation transaction.|F          |1001  |              |1000   |1001|        |0  |        |0  |1001|1001|2  |
|                                      |           |      |              |1000   |1001|        |0  |        |0  |1001|1001|2  |
|Setup: 6. Increment i                 |           |      |              |1000   |1001|1       |1  |        |0  |1003|1001|2  |
|Setup: 7. Build and sign C_i          |C_1        |1004  |              |1000   |1001|        |1  |        |0  |1003|1001|2  |
|Setup: 8. Build and sign D_i          |D_1        |1003  |              |1000   |1001|        |1  |        |0  |1003|1001|2  |
|                                      |           |      |              |1000   |1001|        |1  |        |0  |1003|1001|2  |
|Setup: 10. Submit F                   |           |      |1001          |1001   |1001|        |1  |        |0  |1003|1001|2  |
|                                      |           |      |              |1001   |1001|        |1  |        |0  |1003|1001|2  |
|------                                |           |      |              |1001   |1001|        |1  |        |0  |1003|1001|2  |
|                                      |           |      |              |1001   |1001|        |1  |        |0  |1003|1001|2  |
|Update: Increment i                   |           |      |              |1001   |1001|2       |2  |        |0  |1005|1001|2  |
|Update: Build and sign C_i            |C_2        |1006  |              |1001   |1001|        |2  |        |0  |1005|1001|2  |
|Update: Build and sign D_i            |D_2        |1005  |              |1001   |1001|        |2  |        |0  |1005|1001|2  |
|                                      |           |      |              |1001   |1001|        |2  |        |0  |1005|1001|2  |
|------                                |           |      |              |1001   |1001|        |2  |        |0  |1005|1001|2  |
|                                      |           |      |              |1001   |1001|        |2  |        |0  |1005|1001|2  |
|Update: Increment i                   |           |      |              |1001   |1001|3       |3  |        |0  |1007|1001|2  |
|Update: Build and sign C_i            |C_3        |1008  |              |1001   |1001|        |3  |        |0  |1007|1001|2  |
|Update: Build and sign D_i            |D_3        |1007  |              |1001   |1001|        |3  |        |0  |1007|1001|2  |
|                                      |           |      |              |1001   |1001|        |3  |        |0  |1007|1001|2  |
|---                                   |           |      |              |1001   |1001|        |3  |        |0  |1007|1001|2  |
|                                      |           |      |              |1001   |1001|        |3  |        |0  |1007|1001|2  |
|Close: Submit D_2                     |           |      |1005          |1005   |1001|        |3  |2       |2  |1007|1005|2  |
|Close: Submit D_3                     |           |      |1007          |1007   |1001|        |3  |3       |3  |1007|1007|2  |
|Close: Submit C_3                     |           |      |1008          |1008   |1001|        |3  |        |3  |1007|1007|2  |
|                                      |           |      |              |1008   |1001|        |3  |        |3  |1007|1007|2  |

Ref: https://docs.google.com/spreadsheets/d/1cOXL3mYdMTNDu0FE8lgNMa0fjJ2T-S2jolLG5Dl0xRM

Reviewer 2: yup that answered it thx. Also in case someone else sees this I was wondering about s_e after the close in case _I_ wanted to continue the channel (I was thinking i and e would remain at 3). But after close the variables reset to 0, so s_e after close doesn't matter (as you said)

Author: That's a great point. This is an edge case we should make clearer. I added a section about it in ab21927.

Control is returned to I at close.

- V, the _reserve account_, that holds an amount of native asset contributed
by R that will be used to sponsor the claimable balance ledger entries
created at disbursement.  Created by R.  Jointly controlled by I and R while
the channel is open.  Control is returned to R at close.  Cannot be merged
until all claimable balances created at close for R are claimed by R.

### Constants

The two participants agree on the following constants:

- m, the _maximum transaction count for an iteration's transaction set_, is
defined as 2, the maximum number of transactions that can be signed in any
process between the increments of iteration number i.

### Variables

The two participants maintain the following variables during the lifetime of
the channel:

- s, the _starting sequence number_, is initialized to one greater than the
sequence number of escrow account E after E has been created. It is the first
available sequence number for iterations to consume.

- i, the _iteration number_, is initialized to zero.  It is incremented with
every off-chain update of the payment channel state, or on-chain setup, deposit,
withdrawal, etc.

> - e, the _executed iteration number_, is initialized to zero. It is incremented

Reviewer 2: ❓ I'm still not wrapping my head around why we need a formal variable e. It only increments when the payment channel is closed, at which point it doesn't matter

And therefore s_e is only either s, or the sequence # that E ends up at at the close

Author: e is set to the i of any transaction that the parties agree to submit to the network, so it is used in all the cases where the parties agree to modify the channel after setup, such as withdrawing.

Reviewer 2: got it. I'm wondering if we can modify e's definition then, I think that's where my confusion stemmed from. What about

```
e, the executed iteration number, is initially set to zero, and thereafter represents the i of any transaction set the parties agree to submit to the network.
```

because it changes when the agreement happens. Not after an iteration is submitted on-chain (eg. in the deposit by R steps it changes before anything is submitted)

Reviewer 2: also, if I'm understanding this right I believe P_i should be bumping E.seqNum to s_e, not s_i. [In which case this line](https://github.com/stellar/experimental-payment-channels/pull/3/files#diff-b01917f0debe4bb9e19259c6e34888dd6c41d4f2da28f49af7fc820ba5a38dc7R478) would be 
```
One BUMP_SEQUENCE operation bumping the sequence number of escrow account E to s_e.
```

Author: Ah good point, the description doesn't really say what was intended. Updated in a0b735f. Thank you!

Author: > also, if I'm understanding this right I believe P_i should be bumping E.seqNum to s_e, not s_i. [In which case this line](https://github.com/stellar/experimental-payment-channels/pull/3/files#diff-b01917f0debe4bb9e19259c6e34888dd6c41d4f2da28f49af7fc820ba5a38dc7R478) (and the other changing channel setup cases) would be

I think they're the same thing essentially. In the deposit flow in step 2 when P_i is built s_i will be the sequence number of the escrow account we want to bump to. s_e is not currently set to the value we want to bump to, but it will be after in step 4.

Reviewer 2: ah good point. When I read the construction of P_i it was after going through the steps. At the point P_i is created in your steps s_i is right 👍

with every iteration submitted on-chain.

### Computed Values

The two participants frequently use the following computed values:

- s_i, the _iteration sequence number_, is the sequence number that iteration
i's transaction set starts at. Assuming the history of the payment channel has a
single value for m it is computable as, s+(m*i).

- s_e, the _executed iteration sequence number_, is the sequence number that the
executed iteration e's transaction set starts at. Assuming the history of the
payment channel has a single value for m it is computable as, s+(m*e).

### Processes

#### Setup

To setup the payment channel:

1. I creates the escrow account E.
2. R creates the reserve account V.
3. Set variable initial states:
   - s to E's sequence number + 1.
   - i to 0.
   - e to 0.
5. I and R build the formation transaction F.
6. Increment i.
7. Sign and exchange a closing transaction C_i, that closes the channel with
disbursements matching the initial contributions.
8. Sign and exchange a declaration transaction D_i.

> 9. I and R sign and exchange signatures for formation transaction F.

Reviewer 2: 💡 this wasn't initially obvious to me, wondering if it' worth adding a note about it
```suggestion
9. I and R sign and exchange signatures for formation transaction F.
    - it's important F is signed/exchanged after C_i and D_i in the case I or R disappear
```

Author: Good call, I've added three new sections discussing this at relevant points in the doc. 86a02d0

to R. If I makes a payment of $2 to R, this update process will involve I and R
agreeing on a new declaration and aclosing transaction that supersedes all
previous declaration and closing transaction and that will disburse $8 to I and
$22 to R.

To update the payment channel state, the participants:

1. Increment i.
2. Sign and exchange a closing transaction C_i.
3. Sign and exchange a declaration transaction D_i.

The transactions are constructed as follows:

- C_i, the _closing transaction_, disburses funds to R and changes the signing
weights on E such that I unilaterally controls E.  C_i has source account E,
sequence number s_i+1, a `minSeqAge` of O (the observation period time
duration), and a `minSeqLedgerGap` of O (the observation period ledger count).

  The `minSeqAge` and `minSeqLedgerGap` prevents a misbehaving party from
  executing C_i when the channel state has already progressed to a later
  iteration number, as the other party has the period O to invalidate C_i by
  submitting D_i' for some i' > i.
  
  C_i contains operations:
  - One `BEGIN_SPONSORING_FUTURE_RESERVES` operation that specifies reserve
  account V as a sponsor of future reserves.

>   - One `CREATE_CLAIMABLE_BALANCE` operation for each trustline that is

Reviewer 2: 💡 it depends on if we go with the 2 escrow account idea from your [PR](https://github.com/leighmcculloch/stellar--experimental-payment-channels/pull/1), but if not then I think it's worth adding a point as to why this is a claimable balance op and not a payment op (to prevent a failed C_i)

maybe could be added in the `## Security Concerns` section instead

Author: Good call, I added a paragraph in the security section where the closing transaction failure is discussed. This really relates to that because the use of a payment could cause the closing transaction to fail. a8ff8e7

  disbursing funds to R.
  - One `END_SPONSORING_FUTURE_RESERVES` operation that confirms reserve
  account V's sponsorship.
  - One or more `SET_OPTIONS` operation adjusting escrow account E's
  thresholds to give I full control of E, and removing R's signers.
  - One or more `SET_OPTIONS` operation adjusting reserve account E's
  thresholds to give R full control of V, and removing I's signers.

- D_i, the _declaration transaction_, declares an intent to execute the
corresponding closing transaction C_i.  D_i has source account E, sequence
number s_i, and `minSeqNum` set to s_e.  Hence, D_i can execute at any time, so

> long as E's sequence number n satisfies s_e <= n < s_i.  Because C_i has source

Reviewer 2: ❓ going off my s_e question, could't this just be
```suggestion
long as E's sequence number n satisfies s <= n < s_i.  Because C_i has source
```

Author: When e is zero, then s_e has the same value as s, but when e is greater than zero, it will have a greater value.

is a malicious actor might attempt to do this if an earlier state benefits them.

The malicious participant can do this by performing the [Uncoordinated
Close](#Uncoordinated-Close) process with a declaration transaction that is not
the most recently signed declaration transaction.

The other participant can identify that the close process has started at an
earlier state by monitoring changes in escrow account E's sequence. If the other
participant sees the sequence number of escrow account E change to a value that
is not the most recently used s_i, they can use the following process to contest
the close. A participant contests a close by submitting a more recent
declaration transaction and closing the channel at the actual final state. A
more recent declaration transaction may be submitted because it has a higher
sequence number than the declaration transaction that the malicious actor
submitted. The more recent declaration transaction prevents the malicious actor
from submitting the older closing transaction because it has a lower sequence
number making that transaction invalid.

1. Get E's sequence number n

> 2. If s_{e+1} >= n < s_i, go to step 3, else go to step 1

Reviewer 2: would it be easier if we just said `2. If n != s_i go to step 3 because n is at a malicious value`, implying n is not at the value the latest D_i.seqNum?

the else statement of step 2 says go back to step 1 and re-get n. When would it differ?

Author: For the majority of the channel lifespan n will not be equal to s_i because the escrow account's sequence number will be far in the past compared to the sequence in the last iteration. That's why we need it to be a range. It's okay if the sequence number isn't s_i, as long as it isn't earlier than s_i _and_ later than the last agreed sequence to execute.
